File: game_manager.py
import pieces
from game_state import GameState

class GameManager:
    def __init__(self, gameboard):
        self.gameboard = gameboard
        self.state = GameState.START
        self.scroll_factor = 1 # Wait this many frames between scrolls
        self.frames_since_scroll = 0

    def transition_state(self, new_state):
        if new_state == GameState.MAIN_TITLE:
            self.load_main_title()

    def start(self):
        self.transition_state(GameState.MAIN_TITLE)

    def load_main_title(self):
        # Show title
        title = pieces.MainTitle()
        start_y = int(self.gameboard.board_height - len(title.shape)) # start at bottom
        start_x = int((self.gameboard.board_width - 1) / 2 - len(title.shape[0]) / 2) # centered
        self.gameboard.place_piece(title, start_y, start_x)
        self.scroll_factor = 5

    def update(self):
        if self.frames_since_scroll >= self.scroll_factor:
            self.gameboard.scroll()
            self.frames_since_scroll = 0
        else:
            self.frames_since_scroll += 1

File: src/game-logic/game-manager.js
import pieces from '@/game-logic/pieces'

const gameState = {
  INIT: 'INIT',
  MAIN_TITLE: 'MAIN_TITLE',
  START_PROMPT: 'START_PROMPT',
  GAME: 'GAME'
}

class GameManager {
  constructor (gameboard, playerClass) {
    this.gameboard = gameboard
    this.playerClass = playerClass
    this.player = null
    this.state = gameState.INIT
    this.scrollFactor = 1 // wait this many frames between scrolls
    this.framesSinceScroll = 0
    this.level = 0
  }

  loadMainTitle () {
    // show the title scrolling up the screen
    this.state = gameState.MAIN_TITLE
    let title = new pieces.MainTitle()
    let startY = this.gameboard.boardHeight - title.shape.length // start at bottom of screen
    let startX = Math.round(this.gameboard.boardWidth / 2 - title.shape[0].length / 2)
    this.gameboard.placePiece(title, startY, startX)
    this.scrollFactor = 3
  }

  update (input) {
    // called from main game loop
    if (this.state === gameState.START_PROMPT && input.space.isPressed) {
      this.startGame()
    } else if (this.state === gameState.GAME) {
      this.player.update(input)
    }

    // scroll the board
    if (this.framesSinceScroll >= this.scrollFactor) {
      if (this.state === gameState.GAME) {
        // move the player foward as the game scrolls
        this.player.scroll()
      }
      this.gameboard.scroll(this)
      this.framesSinceScroll = 0
    } else {
      this.framesSinceScroll++
    }
  }

  showStartPrompt () {
    this.state = gameState.START_PROMPT
    this.startPrompt = new pieces.StartPrompt()
    this.startPromptY = Math.round(this.gameboard.boardHeight / 2 - this.startPrompt.shape.length / 2) // center vertically
    this.startPromptX = Math.round(this.gameboard.boardWidth / 2 - this.startPrompt.shape[0].length / 2) // certer horizontally
    this.gameboard.placePiece(this.startPrompt, startPromptY, startPromptX)
    this.scrollFactor = Number.POSITIVE_INFINITY
  }

  startGame (input) {
    // remove start prompt piece
    this.gameboard.removePieceAt(this.startPrompt, startPromptY, startPromptX)
    this.state = gameState.GAME
    this.level = 1
  }

  createPlayer () {
    this.player = new playerClass(this, this.gameboard)
  }

}

export default GameManager